### PR TITLE
[MAGMA][CUDA] cholesky_inverse: deprecate MAGMA and dispatch to cuSolver unconditionally

### PR DESCRIPTION
**Pull request by gderossi**
Both cuSolver and hipSolver implement potrs, so this removes the MAGMA path entirely and adds a deprecation warning, and updates the tests to skip only if missing cuSolver. Thanks @nikitaved for implementing relevant performance improvements in #175898- this PR takes advantage of the triangular solve path as well and benchmarking results are much better than before.

Benchmarking script:
```python
import torch
import torch.utils.benchmark as benchmark

from itertools import product

results = []

batches = [(), (16,), (64,)]
sizes = [16, 128, 512,  2048]

for b, n in product(batches, sizes):
    shape = b + (n, n)
    print(f"Testing shape={shape}")
    label = "torch.cholesky_inverse"
    sub_label = f"{shape}"
    X = torch.rand(*shape, device="cuda")
    X = X @ X.mT + torch.eye(n, device="cuda")
    L = torch.linalg.cholesky(X)
    stmt = "torch.cholesky_inverse(L)"
    for backend in ("magma", "cusolver"):
        torch.backends.cuda.preferred_linalg_library(backend)
        # warm-up
        for _ in range(5):
            exec(stmt)

        results.append(benchmark.Timer(
            stmt=stmt,
            globals={'L': L},
            label=label,
            sub_label=sub_label,
            description=backend,
        ).blocked_autorange(min_run_time=1))

compare = benchmark.Compare(results)
compare.print()
```

Benchmarking results on RTX Pro 6000:
```
[----------- torch.cholesky_inverse ----------]
                        |   magma   |  cusolver | speedup
1 threads: ------------------------------------ |
      (16, 16)          |     37.2  |     31.4  | 1.2
      (128, 128)        |    463.7  |     66.9  | 6.9
      (512, 512)        |   1632.4  |    283.8  | 5.8
      (2048, 2048)      |   7354.0  |   1779.6  | 4.1
      (16, 16, 16)      |     39.1  |     57.8  | 0.7
      (16, 128, 128)    |    597.9  |    159.7  | 3.7
      (16, 512, 512)    |   2947.6  |    852.0  | 3.5
      (16, 2048, 2048)  |  21657.2  |  12394.5  | 1.7
      (64, 16, 16)      |     39.9  |     58.1  | 0.7
      (64, 128, 128)    |    629.5  |    179.2  | 3.5
      (64, 512, 512)    |   4267.9  |   2095.4  | 2.0
      (64, 2048, 2048)  |  63178.9  |  44213.6  | 1.4

Times are in microseconds (us).
```

cc @nikitaved @eqy

### CODE REVIEW

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -887,81 +887,63 @@ REGISTER_CUDA_DISPATCH(cholesky_stub, &cholesky_kernel)
...
-      } else {
-        return cholesky_inverse_kernel_impl_magma(result, infos, upper);
-      }
-  }
-#else
-  return cholesky_inverse_kernel_impl_magma(result, infos, upper);
-#endif
-
+  _warn_once_magma_deprecation("linalg.cholesky_inverse");
+  return _cholesky_inverse_helper_cuda_cusolver_dispatcher(result, infos, upper);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: We do not need the helper functions -- we can directly dispatch to `cholesky_solve` with the RHS being the identity.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -887,81 +887,63 @@ REGISTER_CUDA_DISPATCH(cholesky_stub, &cholesky_kernel)
...
+        const auto L = upper
+          ? c10::MaybeOwned<Tensor>::owned(input_working_copy.mH())
+          : c10::MaybeOwned<Tensor>::borrowed(input_working_copy);
+        // NOTE: we tolerate redispatch with at::triangular_solve_triangular
+        // because it handles memory layout optimization and conj/neg flags.
+        // IMPORTANT NOTE: `self` and `A` are not processed for kernel calls yet!
+        // Step 1: Solve for Y: L Y = B or U^H Y = B.
+        auto X = at::linalg_solve_triangular(*L, result, /*upper=*/false);
+        // Step 2: Solve for X: L^H X = Y or U X = Y.
+        at::linalg_solve_triangular_out(result, L->mH(), X, /*upper=*/true);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: OK, fine with having a helper function, but the first system should be solved in-place wrt `result`, and so is the second one.

We can still remove code duplicate by modifying `cholesky_solve` logic to perform two in-place solves (these override RHS), and then by dispatching to it in both methods (i.e. solve and inverse) with different system matrices/RHSs. This is up to you, I will not block on this.
